sql/catalyst/src/main/java/org/apache/spark/sql/catalog/v2/TableCatalog.java

```diff
@@ -0,0 +1,153 @@
+ * the License.  You may obtain a copy of the License at
+ *
+ *    http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package org.apache.spark.sql.catalog.v2;
+
+import org.apache.spark.sql.catalog.v2.expressions.Transform;
+import org.apache.spark.sql.catalyst.analysis.NoSuchNamespaceException;
+import org.apache.spark.sql.catalyst.analysis.NoSuchTableException;
+import org.apache.spark.sql.catalyst.analysis.TableAlreadyExistsException;
+import org.apache.spark.sql.sources.v2.Table;
+import org.apache.spark.sql.types.StructType;
+
+import java.util.Collections;
+import java.util.Map;
+
+/**
+ * Catalog methods for working with Tables.
+ */
+public interface TableCatalog extends CatalogPlugin {
+  /**
+   * List the tables in a namespace from the catalog.
+   *
+   * @param namespace a multi-part namespace
+   * @return an array of Identifiers for tables
+   * @throws NoSuchNamespaceException If the namespace does not exist (optional).
+   */
+  Identifier[] listTables(String[] namespace) throws NoSuchNamespaceException;
+
+  /**
+   * Load table metadata by {@link Identifier identifier} from the catalog.
+   *
+   * @param ident a table identifier
+   * @return the table's metadata
+   * @throws NoSuchTableException If the table doesn't exist.
+   */
+  Table loadTable(Identifier ident) throws NoSuchTableException;
+
+  /**
+   * Refresh table metadata for {@link Identifier identifier}.
+   * <p>
+   * If the table is already loaded or cached, drop cached data and reload it. If the table is not
+   * already loaded, load it and update caches with the new version.
+   *
+   * @param ident a table identifier
+   * @return the table's current metadata
+   * @throws NoSuchTableException If the table doesn't exist.
+   */
+  default Table refreshTable(Identifier ident) throws NoSuchTableException {
+    return loadTable(ident);
+  }
+
+  /**
+   * Test whether a table exists using an {@link Identifier identifier} from the catalog.
+   *
+   * @param ident a table identifier
+   * @return true if the table exists, false otherwise
+   */
+  default boolean tableExists(Identifier ident) {
+    try {
+      return loadTable(ident) != null;
+    } catch (NoSuchTableException e) {
+      return false;
+    }
+  }
+
+  /**
+   * Create a table in the catalog.
+   *
+   * @param ident a table identifier
+   * @param schema the schema of the new table, as a struct type
+   * @return metadata for the new table
+   * @throws TableAlreadyExistsException If a table already exists for the identifier
+   * @throws NoSuchNamespaceException If the identifier namespace does not exist (optional)
+   */
+  default Table createTable(
```

Comment:
This is implemented by users and called by Spark, which makes me think that overload doesn't make the API better.

There are 3 `createTable`, and users can override all of them, which is not expected. I think we should make API as strict as possible, and only have one `createTable`.

On the other hand, it's called by Spark so it doesn't hurt to write down the default values at the caller side.

Reply:
I think this is fine because there are default methods. Users have the option to implement these if they choose, but can implement just the last one if they want. That allows them flexibility and is friendlier for Spark callers. I don't think there is a strong drawback to having this so I'd rather keep it in.

Reply:
The drawback I can think of is that we can't guarantee the default values if the implementation overrides these overload methods.

Reply:
If the implementation overrides these methods, then it can choose how to default these or whether it needs to default them.

While I don't think your argument holds, I'd rather get this in than have these methods. I'll remove them.